### PR TITLE
Added laser data from map in nav2_loopback_sim

### PR DESCRIPTION
**Pull request by JatinPatil2003**
---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | #4616 |
| Primary OS tested on | Humble |
| Robotic platform tested on |  |
| Does this PR contain AI generated software? | (No; Yes and it is marked inline in the code) |

---

## Description of contribution in a few bullet points

* Publishes realistic laser data according to the map

---

#### For Maintainers: 
- [ ] Check that any new parameters added are updated in docs.nav2.org
- [ ] Check that any significant change is added to the migration guide
- [ ] Check that any new features **OR** changes to existing behaviors are reflected in the tuning guide
- [ ] Check that any new functions have Doxygen added
- [ ] Check that any new features have test coverage
- [ ] Check that any new plugins is added to the plugins page
- [ ] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists

### CODE REVIEW

File: nav2_loopback_sim/nav2_loopback_sim/utils.py
Comment (SteveMacenski) [inline nav2_loopback_sim/nav2_loopback_sim/utils.py:70]: Use the bound checks in https://github.com/ros-navigation/navigation2/blob/3f311124e356dbfa33f35e23caa5474a57f99ab7/nav2_costmap_2d/src/costmap_2d.cpp#L284-L297

File: nav2_loopback_sim/nav2_loopback_sim/utils.py
Comment (SteveMacenski) [inline nav2_loopback_sim/nav2_loopback_sim/utils.py:76]: This appears to me to incorrect? https://github.com/ros-navigation/navigation2/blob/3f311124e356dbfa33f35e23caa5474a57f99ab7/nav2_costmap_2d/src/costmap_2d.cpp#L278-L282

File: nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py
Comment (SteveMacenski) [inline nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py:32]: Rename them to be inline with the camelCase naming convention please :-)

File: nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py
Comment (SteveMacenski) [inline nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py:105]: For all: use camelcase for naming things `getMap()`

File: nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py
Comment (SteveMacenski) [inline nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py:189]:
```suggestion
        # self.scan_msg.header.stamp = (self.get_clock().now()).to_msg()
```

File: nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py
Comment (SteveMacenski) [inline nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py:260]:
```suggestion
                'map', self.scan_frame_id, now, rclpy.duration.Duration(seconds=0.1))
```

File: nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py
Comment (SteveMacenski) [inline nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py:297]: I think if the robot is off the map, we can just set it to max value -- but I'm not totally grokking this. can you explain?

File: nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py
Comment (SteveMacenski) [inline nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py:241]: I think this log should mention that the laser scan will be populated

File: nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py
Comment (SteveMacenski) [inline nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py:244]: These should fail by saying that the max range will be used instead

File: nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py
Comment (SteveMacenski) [inline nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py:235]:
```suggestion
        if self.map_client.wait_for_service(timeout_sec=5.0):
```

if we're only trying once, we should try hard

File: nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py
Comment (SteveMacenski) [inline nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py:254]: Remove this, I'd try once on startup and after that just use max. This adds alot of latency

File: nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py
Comment (SteveMacenski) [inline nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py:281]: or initial pose not set

Comment (SteveMacenski) [inline nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py:281]: This `if` shouldn't be in the loop. It should be checked at the start (with the initial pose addition) and then returned immediately with the max range minus 0.1

File: nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py
Comment (SteveMacenski) [inline nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py:313]: I'm not really grokking the voxel_border stuff from L313 to L337. Can you explain? I think the worldToMap and mapToWorld conversions should be handling the `0.5`'s fine itself. I don't get the other stuff

File: nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py
Comment (SteveMacenski) [inline nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py:288]: This function has some analogs to https://github.com/ros-navigation/navigation2/blob/main/nav2_simple_commander/nav2_simple_commander/line_iterator.py. Perhaps this can be used instead to reduce code duplication and potential errors for ray marching? Example use here: https://github.com/ros-navigation/navigation2/blob/main/nav2_simple_commander/nav2_simple_commander/footprint_collision_checker.py#L113-L126 you could march from start -> max range and keep going until you hit an occupied cell, max range, or leave the map

**Review by SteveMacenski — CHANGES_REQUESTED** [review]
Thanks for taking this on! Largely looks great but a few bugs / nits that are good to have resolved that I think makes it even better :-)

File: nav2_loopback_sim/nav2_loopback_sim/utils.py
Comment (SteveMacenski) [reply nav2_loopback_sim/nav2_loopback_sim/utils.py:76]: We have reference implementations: https://github.com/ros-navigation/navigation2/blob/main/nav2_simple_commander/nav2_simple_commander/costmap_2d.py#L143-L186

File: nav2_loopback_sim/nav2_loopback_sim/utils.py
Comment (JatinPatil2003) [reply nav2_loopback_sim/nav2_loopback_sim/utils.py]: Updated

File: nav2_loopback_sim/nav2_loopback_sim/utils.py
Comment (JatinPatil2003) [reply nav2_loopback_sim/nav2_loopback_sim/utils.py]: Both were same but I have changed it to be same as reference

File: nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py
Comment (JatinPatil2003) [reply nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py]: As `worldToMap` is updated to check bounds, I have removed this condition

Originally it was implemented [here](https://github.com/nobleo/mobile_robot_simulator/blob/master/src/laser_simulator.cpp) which will keep populating laser scan data even when the robot gets out of the map

File: nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py
Comment (JatinPatil2003) [reply nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py]: I have tried to keep it same as original [reference](https://github.com/nobleo/mobile_robot_simulator/blob/master/src/laser_simulator.cpp)

File: nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py
Comment (JatinPatil2003) [reply nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py]: okay i'll change logic using `LineIterator` to find laser ranges

**Comment by JatinPatil2003** [comment]
Converting PR to draft for implementation of `LineIterator`

File: nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py
Comment (SteveMacenski) [inline nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py:241]:
```suggestion
                self.get_logger().info('Laser scan will be populated using map data')
```

File: nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py
Comment (SteveMacenski) [inline nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py:107]: Move before the initialized log

File: nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py
Comment (SteveMacenski) [inline nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py:243]: All part of 1 log (ditto below)

File: nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py
Comment (SteveMacenski) [inline nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py:252]: `None`s? (as well in the exception handling) That would make it the same as the map to world conversion methods error handling

File: nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py
Comment (SteveMacenski) [inline nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py:289]: Should this be `ranges.append()` instead?

**Comment by SteveMacenski** [comment]
Thanks! Looks so much better already! :-)

File: nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py
Comment (JatinPatil2003) [reply nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py:289]: both can be done actually I have initialized `self.scan_msg.ranges = [0.0] * num_samples` already and in earlier commits it was not returning anything when the scan range is less than min range do by default it was `0.0`

but later I have changed it to return `0.0`, So now `.append()` will also work

but let me know should I change it to `.append()`

**Comment by JatinPatil2003** [comment]
> Thanks! Looks so much better already! :-)

@SteveMacenski Should I change PR status then?

File: nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py
Comment (SteveMacenski) [reply nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py:293]: Oh ok! That works, I missed that

**Comment by SteveMacenski** [comment]
I assume you tested this and see a reasonable laser scan that Nav2 can use? Just checking before I do my final pass through that 'this has shown to work' :-)

**Comment by JatinPatil2003** [comment]
https://github.com/user-attachments/assets/99a60754-9fe5-469e-bfdf-044b5f6246b2

This is working of nav2_loopback_sim, laser scan with nav2

File: nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py
Comment (JatinPatil2003) [reply nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py:141]: I did try publishing laser scan on same timer but as ray tracing is more instruction heavy it is taking more time to populate laser scan and due to that odometry and tf performance is getting affected 

maybe if 2 timers are not optimal solution then making a separate node to publish laser scan makes sense

**Comment by JatinPatil2003** [comment]
> I generally approve except for the `findMapRange` stuff which should use the line iterator
> 
> > This is working of nav2_loopback_sim, laser scan with nav2
> 
> Generally looks good! I think something is possibly going on (?) since the laser scan appears to be flickering between 2 solutions, even when the robot is still. These should perfectly line up since there's no source of error. If there's no error/noise and the robot's TF doesn't move, this should be completely deterministic. I'm wondering if that's from something in the ray tracing that the line iterator should resolve or the call to TF. You use TF lookups to get the map->laser frame, but we do have map->odom and odom->base_link transforms in the simulator exactly. I suppose if you looked up at startup base_link->laser frame and cached that, then it should be 100% aligned (if that's the error cause)

Yea it's not due to ray tracing, while testing with nav it tested with nav2_bringup so amcl is also publishing another tf

I think a parameter for publishing map to odom will be the solution for this

File: nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py
Comment (JatinPatil2003) [reply nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py:189]: As I'm using different timers for scan publishing stamp will be needed

File: nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py
Comment (SteveMacenski) [inline nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py:75]: I don't understand this parameter. Why? 

At the very least, it should be default to true

File: nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py
Comment (SteveMacenski) [inline nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py:79]: Make it the duration like we have for the other update parameter

File: nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py
Comment (SteveMacenski) [inline nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py:149]: Then don't need to invert

File: nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py
Comment (SteveMacenski) [reply nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py:197]: Please do not resolve conversations that you did not complete.

**Review by SteveMacenski — CHANGES_REQUESTED** [review]
Did you try the line iterator method? Perhaps you didn't push this to `origin` so I could see

File: nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py
Comment (SteveMacenski) [reply nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py:75]: > I think a parameter for publishing map to odom will be the solution for this

I don't understand the take away. I'm saying that we shouldn't use TF buffer lookups. We have `self.t_map_to_odom` and `self.t_odom_to_base_link`. If trying to find the laser pose, we only need to find on initialization the `self.t_base_link_to_laser_scanner` and we have all the exact transforms needed to find the laser scanner link in the map frame without involving TF

Comment (JatinPatil2003) [reply nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py]: In latest commit I used the cached tf but still the same problem of flickering is there, when this parameter is set to False then it work well with nav2(no flickering)

**Both tested with LineIterator and with using cached tf**

https://github.com/user-attachments/assets/84a90c31-5bb2-49b2-8487-f619b665864b


https://github.com/user-attachments/assets/1f2aa291-0389-46ff-a17e-6c70b325b63e

**Comment by JatinPatil2003** [comment]
I think the current ray tracking algorithm is working well as compared to LineIterator
This case is only when robot is near obstacle but works well when not too close

**This image is using LineIterator**
image

**This Image is using current ray tracking algorithm**
image

File: nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py
Comment (SteveMacenski) [reply nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py]: OK - we can leave as is

File: nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py
Comment (SteveMacenski) [reply nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py]: Is that happening before your changes? You're also showign a particle cloud, so I don't know what you mean by true/false because they're both using AMCL. This replaces AMCL when set to true -- maybe that's why you have flickering. You're running both simultaneously.

Try using the provided loopback launch files I provide in `nav2_bringup` - which does not have AMCL.

File: nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py
Comment (SteveMacenski) [inline nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py:159]: These are non-static, do not store, get when needed. We don't want to have duplicate information we're trying to keep consistent.

File: nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py
Comment (SteveMacenski) [inline nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py:211]: These are non-static, do not store, get when needed. We don't want to have duplicate information we're trying to keep consistent.

File: nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py
Comment (SteveMacenski) [inline nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py:212]: These are non-static, do not store, get when needed. We don't want to have duplicate information we're trying to keep consistent.

File: nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py
Comment (SteveMacenski) [inline nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py:283]: Transform here when needed

**Review by SteveMacenski — CHANGES_REQUESTED** [review]

**Comment by SteveMacenski** [comment]
> This image is using LineIterator

I like this in concept. I'm a little unclear why they seem to "stack up" to so much thickness of the wall -- shouldn't it have hit one of the previous cells before being basically on the other side of the wall's thickness? We do this in the [collision footprint checker](https://github.com/ros-navigation/navigation2/blob/main/nav2_simple_commander/nav2_simple_commander/footprint_collision_checker.py#L113) in cell coordinates with 0.5 cell steps -- maybe that's worth a try instead of using world coordinates? `line_iterator = LineIterator(worldToMap(float(x0)), worldToMapfloat(y0)), worldToMap(x1), worldToMap(y1), 0.5)` 

> This Image is using current ray tracking algorithm

I also like this in concept, but the arcing is a bit unnatural and I think would cause some odd behaviors with some localization algorithms.

**Comment by JatinPatil2003** [comment]
> This image is using LineIterator

This has been fixed. Now it is working perfectly

File: nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py
Comment (SteveMacenski) [inline nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py]: Why is this removed? Please readd

File: nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py
Comment (SteveMacenski) [inline nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py:205]:
```suggestion
        # self.scan_msg.header.stamp = (self.get_clock().now()).to_msg()
```

Please leave this as commented out. I'll test after this is approved to see if readding that still works in the launch file

Comment (SteveMacenski) [reply nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py:205]: Note to self:
- Test localization
- Test SLAM Toolbox
- Test with TB3/4 launch files

File: nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py
Comment (SteveMacenski) [inline nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py:311]: `range_max` should be in grid coordinates ( divided by resolution)

**Review by SteveMacenski — APPROVED** [review]

File: nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py
Comment (JatinPatil2003) [reply nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py:311]: I don't get it x0 and x1 are world coordinates so basically distance between (x0, y0) and (x1, y1) should be range_max, which should be in meters

let me know if I'm wrong here

File: nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py
Comment (SteveMacenski) [inline nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py:209]: OK - I played with this and this can be uncommented out. I'm not sure why I was having issues before.

File: nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py
Comment (SteveMacenski) [inline nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py:218]:
```suggestion
        self.scan_msg.range_max = 30.0
```

File: nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py
Comment (SteveMacenski) [inline nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py:209]:
```suggestion
        self.scan_msg.header.stamp = (self.get_clock().now()).to_msg()
```

**Comment by SteveMacenski** [comment]
Thanks @JatinPatil2003 !

**Pull request merged**
